### DISCUSSION TITLE
Local development workflow is painfully slow.

### DISCUSSION THREAD

**Discussion by akutruff** [supabase/supabase#14598] [Questions]
Maybe I'm doing this wrong but the workflow of making a single change to SQL and applying it to the local database takes forever.

1. Make change to a SQL file
1. `supabase db reset`
1. Wait for 21 seconds while looking at `Restarting containers...`  ( I actually timed that with a stop watch)

Meanwhile, if you have a tool like pgadmin open, you need to renavigate to the table affected to see the changes as the entire database has been wiped and re-inited.  If I try to use Supabase Studio, it's just as bad because I have my data in a different schema than public, and I have to scroll a million schemas in the menu to find.  It's very tedious.

Perhaps I'm doing it all wrong, or is this everyone else's DX as well?

**Comment by akutruff** [comment]
bump. Any comments appreciated, here.  Not sure I can keep using supabase with this DX, and I'd really like to after comparing so many other offerings. :(

**Reply by vickkhera** [reply]
While you're making rapid changes, just apply them to your local copy of the DB my running the migration file manually using the `psql` CLI, or copy/pasting minor changes in function definitions and the like. Once you're satisfied with it, run `supabase db reset` to verify it all works in sequence, then commit it.

**Reply by akutruff** [reply]
Thanks for the reply.

So that means I need to make a migrate down as well?  Seems to defeat the point.

What I don't understand is why they have to restart the containers rather than work from a last know snapshot without latest migrations.

**Comment by pongraczi** [comment]
Try this one:  
https://sqitch.org/  

I used to use this to manage upgrade/downgrade DB, including any changes.  

So, the workflow could be:  
* supabase reset
* starting a new or continue an existing project
* edit the sql files with your favourite editor
* sqitch deploy
* check the result
* revert, fix issues, deploy again

I like it very much.

[Postgresql tutorial](https://sqitch.org/docs/manual/sqitchtutorial/)

**Reply by akutruff** [reply]
I actually switched to supabase hoping to replace my usage of sqitch because it was error prone if there was a mistake in the migrate down files.

**Comment by akutruff** [comment]
So is there no way to clear the DB and run migrations without waiting for the super long container restart?

**Comment by bombillazo** [comment]
I dont understand your work flow. Why are you resetting your DB on every SQL change?

**Reply by akutruff** [reply]
How else will the change show up in my DB?  

I create a new migration.  I create a table in the migration.  `migration up` Oops, I realized I need to change a type of a column.  I have to run `db reset` now.  As I'm iterating I want to just undo and redo the latest migration without having to wait 20 seconds to see the result.

**Reply by bombillazo** [reply]
Why are you running migrations for every single change, though? Migrations are mainly used when you want to commit the snapshot of your changes to source code control. You can quickly iterate locally with various changes, and when they are all done, create a migration file for those related changes. You modify your database via the Supabase Studio GUI, connect to it via a client like DBeaver and run your SQL/also edit via its GUI, or even via Postgres terminal commands. You don't need migrations to change your local DB.

**Reply by bombillazo** [reply]
The only way to "revert" a migration is like most other ORMs do this: They have a "revert/down" migration SQL file that runs the opposite process in the migration. Supabase migrations don't generate those automatically. The other solution is to create an SQL script that clears your local DB and then run `supabase migrations up` (I believe this is the command, not sure if its documented but it exists) to execute all the SQL files in your migrations folder (of course, remove the migration you don't want to have run).

**Reply by pongraczi** [reply]
aaaaand we are back to sqitch :))))))))  
That was the main reason, I started using sqitch.

**Comment by DarkLite1** [comment]
I'm new to supabase and using the same workflow as described in the [docs](https://supabase.com/docs/guides/deployment/database-migrations). The part `supabase db reset` really takes ages. On my laptop this takes about 90 seconds.... Very annoying.

**Comment by fundclaimpro** [comment]
Hi is someone available to assist pls?

**Comment by fabyanMikhael** [comment]
@DarkLite1  @fundclaimpro 

try this:

get your local ip of where your supabase docker is running. for example in wsl you can run `hostname -I` 

and then do this:

`npx supabase db reset --db-url postgresql://postgres:postgres@<local ip>:54322/postgres`

if you use `localhost` or `127.0.0.1` it seems to detect that it's a local supabase environment and tries to reset the containers, which is probably what takes super long.

if you use an ip even if it is your local device's ip, then it will consider it a remote database and only reset the database which is nearly instant...

**Comment by GokulTowner1** [comment]
That slow schema browsing on large DBs pain is real — I hit the exact same wall with pgAdmin on a ~150-table schema and it's what pushed me to try building something from scratch.

The core issue is that most tools (even the Electron-based ones) do schema introspection synchronously or in a single thread. With pgStudio I went with Tauri (Rust backend) specifically so schema loading and query execution run on native threads — it stays responsive even on 300+ table databases.

If you're curious, it's on the Mac App Store: https://apps.apple.com/in/app/pgstudio/id6759511669

That said, I'd actually love to hear more about your setup — how many tables/columns are you working with, and what client are you using now? Even if you don't try pgStudio, knowing the real-world pain points helps me build something better.
